File: haskell/shapes.hs
import Control.Monad (guard)

data Shape
    = Circle Float
    | Rectangle Float Float

area :: Shape -> Float
area (Circle r) = pi * r * r
area (Rectangle h w) = h * w

perimeter :: Shape -> Float
perimeter (Circle r) = 2 * pi * r
perimeter (Rectangle h w) = 2 * (h + w)

main = do
    guard $ area (Circle 10) == (100 * pi)
    guard $ perimeter (Circle 10) == (20 * pi)
    guard $ area (Rectangle 2 8) == 16
    guard $ perimeter (Rectangle 2 8) == 20

File: ts/shapes.ts
import { strict as assert } from "assert"

interface Shape {
  area: () => number
}

class Circle implements Shape {
  constructor(public radius: number) {}
  area() {
    return Math.PI * this.radius ** 2
  }
}

class Rectangle implements Shape {
  constructor(public width: number, public height: number) {}
  area() {
    return this.width * this.height
  }
}

let shapes: Shape[] = [new Circle(10), new Rectangle(5, 10)]
assert(shapes[0].area() === 314.1592653589793)
assert(shapes[1].area() === 50)
